### PR TITLE
Allow open-chain messages to be rejected

### PR DESCRIPTION
**Pull request by ma2bd**
## Motivation

`OpenChain` messages can be created by user operations. After #1475, they could be grouped with user messages in a fallible transaction. Therefore, `OpenChain` messages have to be rejectable.

## Proposal

* Make `OpenChain` messages "simple" instead of "protected".
* Rejecting an `OpenChain` message is allowed and close the chain immediately.
* When closing a chain in the middle of a block, operations or accepted messages are now disallowed right after the chain is closed (not at the next block).
* Empty blocks on closed chains are allowed but they have to pay a block fee.

I didn't change the `open_chain` syscall or the `OpenChain`operation but after #1475, we should probably remove the initial transfer entirely. (This transfer is redundant and restricted to chain-to-chain accounts)

## Test Plan

CI

## Release Plan

- Contains breaking changes.
- Next devnet ok

### CODE REVIEW

File: linera-chain/src/chain.rs [hunk 1]
```diff
@@ -971,9 +978,12 @@ where
...
         }
 
-        // Finally, charge for the block fee, except if the chain is closed. Closed chains should
-        // always be able to reject incoming messages.
-        if !self.is_closed() {
+        // Finally, charge for the block fee, except if the chain is closed and the block
+        // was used to reject some incoming messages (and only that).
+        if !self.is_closed()
+            || block.incoming_messages.is_empty()
+            || !block.has_only_rejected_messages()
```
Comment (afck) [inline linera-chain/src/chain.rs:985]: Is the third line needed?
If it is closed, doesn't that imply that it has only rejected messages?

File: linera-execution/tests/test_execution.rs [hunk 2]
```diff
@@ -1520,18 +1520,35 @@ async fn test_open_chain() {
...
+
     // Initialize the child chain using the config from the message.
     let mut child_view = SystemExecutionState::default()
         .into_view_with(ChainId::child(message_id), Default::default())
         .await;
     child_view
         .system
         .initialize_chain(message_id, Timestamp::from(0), config.clone());
-    assert_eq!(*child_view.system.balance.get(), Amount::ONE);
+    assert_eq!(*child_view.system.balance.get(), Amount::ZERO);
```
Comment (afck) [inline linera-execution/tests/test_execution.rs:1551]: (We could make the child handle the credit message and then assert the balance again, for completeness. But maybe that's redundant.)

File: linera-service/tests/end_to_end_tests.rs [hunk 3]
```diff
@@ -3163,17 +3163,22 @@ async fn test_end_to_end_faucet(config: impl LineraNetConfig) -> Result<()> {
 
     // Clients 2 and 3 should have the tokens, and own the chain.
     client2.sync(chain2).await?;
+    // We haven't process incoming messages yet.
```
Comment (afck) [inline linera-service/tests/end_to_end_tests.rs:3166]:
```suggestion
    // We haven't processed incoming messages yet.
```

File: linera-service/tests/end_to_end_tests.rs [hunk 4]
```diff
@@ -3163,17 +3163,22 @@ async fn test_end_to_end_faucet(config: impl LineraNetConfig) -> Result<()> {
...
     assert_eq!(
         client2.local_balance(Account::chain(chain2)).await?,
-        Amount::from_tokens(2),
+        Amount::from_tokens(0),
```
Comment (afck) [inline linera-service/tests/end_to_end_tests.rs:3169]: (There's also `Amount::ZERO`.)

**Review by afck — APPROVED** [review]

File: linera-chain/src/chain.rs (hunk 1 above)
Comment (ma2bd) [reply linera-chain/src/chain.rs:985]: After this PR, it could have been closed in the middle of the block.

**Comment by ma2bd** [comment]
@afck is working on finishing this

**Pull request closed without merging**